### ISSUE TITLE
Binder appears to write main cpp file before checking if output directory exists

### ISSUE THREAD

**Issue by fake-name** [RosettaCommons/binder#57]
This is a weird edge-case that might be a confluence of a number of things.

Basically, I have a project built as follows

```
#!/bin/bash

rm -rfv /media/sf_Code/scfft/libraries/sndlib_py/generated_binding/
sync

/home/durr/binder/build/llvm-4.0.0/build_4.0.0.linux.desktest.release/bin/binder \
	-v \
	--root-module sndlib_py                                                       \
	--prefix /media/sf_Code/scfft/libraries/sndlib_py/generated_binding/         \
	--config binder_config.config                                                \
	--bind "my_namespace"                                                               \
	binder_includes.hpp                                                          \
	--                                                                           \
	 -std=c++11                                                                  \
	-I/media/sf_Code/scfft                                                       \
	-DNDEBUG
```

If I have the `rm` command remove everything *including* the `generated_binding` directory, the generator runs without error and produces the expected output for everything, but `sndlib_py.cpp` never gets written. However, if I change the `rm` command to not delete the directory:

```
rm -rfv /media/sf_Code/scfft/libraries/sndlib_py/generated_binding/*
```
 `sndlib_py.cpp`  gets successfully written.

There's a bunch of confounding factors, but it appears that the writing of at least *some* of the output files is not properly checked for success. Additionally, verbose mode doesn't ever mention the writing of  `sndlib_py.cpp`, though it does mention all the other generated `*.cpp` files.

I'm working in a virtualbox VM which is mounting a host directory, so there could be shared folder oddness going on, but the issue seems to be quite reproducible.

**Comment by fake-name** [comment]
Ok, I moved into a plain folder on the local drive, and can confirm this exists without all the network/shared folder oddments.

Basically, if the path specified by `--prefix` doesn't exist, it is created, but `{root-module}.cpp` is never created.

**Comment by lyskov** [comment]
@fake-name thank you for reporting this! The prefix directory is indeed should be present before run so such failure is expected, however we should indeed report an error in such cases. I will looks this up.

**Comment by fake-name** [comment]
The prefix directory is created during the run, but I think there's a race condition, or a mis-ordering so it's created after the attempt to write the module root. The issue is repeatable, at least for me.

**Comment by HFTrader** [comment]
I struggle with this as well until I realized the problem. It's interesting this issue has been hanging for a couple of years, seems easy enough to me to solve. So either:
1. Check if the prefix folder exists and bail out if not; or
2. At start, create the prefix folder if it does not exist, recursively
Of course, now that I spent hours on this, I know and I've put a `mkdir -p <prefix>` on my script, for sanity.
But it should indeed be solved within binder, for the mental sake of new users.
